File: challenges-in-python/k_way_merge/merge_sorted_array.py
# Solving the "Merge Sorted Array" Challenge
# Challenge Description:
# Given two sorted arrays nums1 and nums2, merge nums2 into nums1 in a sorted manner. 
# Assume that nums1 has enough space to hold additional elements from nums2.

# Approach:
# A straightforward approach would involve inserting all elements from nums2 into nums1 and then sorting, 
# but this wouldn't be efficient. A better approach is to fill nums1 from the end, 
# comparing elements from nums1 and nums2 and placing the larger ones first. 
# This way, we avoid overwriting elements in nums1 that haven't been checked yet.

def merge(nums1, m, nums2, n):
    # Set pointers to the end of each array
    p1 = m - 1
    p2 = n - 1
    # Set pointer for the end of the merged array
    p = m + n - 1

    # While there are elements in nums1 and nums2
    while p2 >=0 :
        # If there are elements in nums1 and nums2, compare them
        if p1 >= 0 and nums1[p1] > nums2[p2]:
            nums1[p] = nums1[p1]
            p1 -= 1
        else:
            nums1[p] = nums2[p2]
            p2 -= 1
        p -= 1

# Test cases
def test_merge():
    nums1 = [1, 2, 3, 0, 0, 0]
    m = 3
    nums2 = [2, 5, 6]
    n = 3
    merge(nums1, m, nums2, n)
    assert nums1 == [1, 2, 2, 3, 5, 6]

    nums1 = [0]
    m = 0
    nums2 = [1]
    n = 1
    merge(nums1, m, nums2, n)
    assert nums1 == [1]


test_merge()

File: challenges-in-rust/src/k_way_merge/merge_sorted_array.rs
// Solving the "Merge Sorted Array" Challenge
// Problem Statement
// Given two sorted arrays nums1 and nums2, 
// merge nums2 into nums1 in-place, maintaining the sorted order. 
// Assume that nums1 has enough space to hold additional elements from nums2.

// Implementation
// This problem can be solved efficiently by using a two-pointer approach starting from the end of both arrays, 
// which aligns with the K-way merge pattern when K=2.

// The merge function takes two mutable references to the input arrays nums1 and nums2,
// and the lengths of the two arrays as arguments.

// Big O Analysis:
// time complexity: O(m + n) | space complexity: O(1)
pub fn merge(nums1: &mut Vec<i32>, m: i32, nums2: &mut Vec<i32>, n: i32) {
    let (mut p1, mut p2, mut p) = (m - 1, n - 1, (m + n - 1) as usize);
    while p2 >= 0 {
        if p1 >= 0 && nums1[p1 as usize] > nums2[p2 as usize] {
            nums1[p] = nums1[p1 as usize];
            p1 -= 1;
        } else {
            nums1[p] = nums2[p2 as usize];
            p2 -= 1;
        }
        if p >= 1 { // to avoid overflow
            p -= 1;
        }
    }
}

// Let's test the merge function with a few test cases:
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_merge() {
        let mut nums1 = vec![1, 2, 3, 0, 0, 0];
        let m = 3;
        let mut nums2 = vec![2, 5, 6];
        let n = 3;
        merge(&mut nums1, m, &mut nums2, n);
        assert_eq!(nums1, vec![1, 2, 2, 3, 5, 6]);

        let mut nums1 = vec![1];
        let m = 1;
        let mut nums2 = vec![];
        let n = 0;
        merge(&mut nums1, m, &mut nums2, n);
        assert_eq!(nums1, vec![1]);
        
        let mut nums1 = vec![0, 0, 0];
        let m = 0;  
        let mut nums2 = vec![1, 2, 3];
        let n = 3;
        merge(&mut nums1, m, &mut nums2, n);
        assert_eq!(nums1, vec![1, 2, 3]);
    }
}